tests/test_pipeline.py

```diff
@@ -32,6 +49,159 @@
 
 _RANDOM_STATE = 67056
 
+# This is a global variable that is used to count the number of fit_transformations
```

Comment:
I see the problem with the counter variable. 

1. Do not use global variables here.
2. Test for correctness in `test_caching` and `test_gridseach_cache`
3. Test number of executions (counter) only in `test_caching` because GridSearchCV always clones the estimator, also when n_jobs=1, making it impossible to count the `fit_transform` executations.
4. Move the helper classes `CountingTransformerWrapper` into a utils file.

Reply:
Done